File: frontend/src/types/teachingLoad.ts
// Teaching load (Нагрузка) wire types. Mirror of the backend DTO in
// internal/modules/schedule/application/dto/teaching_load_dto.go and the
// TeachingLoad entity. The auto-scheduler (issue #139) expands each load line
// into schedulable lessons; this screen is the methodist's source of truth.

import type { WeekType } from './schedule'

export type { WeekType }

// TeachingLoad is one planned assignment with denormalized reference names.
export interface TeachingLoad {
  id: number
  semester_id: number
  group_id: number
  group_name: string
  discipline_id: number
  discipline_name: string
  teacher_id: number
  teacher_name: string
  lesson_type_id: number
  lesson_type_name: string
  pairs_per_week: number
  week_type: WeekType
}

// TeachingLoadInput is the create/update request body (7 mutable fields).
export interface TeachingLoadInput {
  semester_id: number
  group_id: number
  discipline_id: number
  teacher_id: number
  lesson_type_id: number
  pairs_per_week: number
  week_type: WeekType
}

// TeachingLoadFilter narrows the list; all fields optional.
export interface TeachingLoadFilter {
  semester_id?: number
  group_id?: number
  teacher_id?: number
}

// TeachingLoadListResponse is the wrapped list payload.
export interface TeachingLoadListResponse {
  teaching_loads: TeachingLoad[]
}

File: internal/modules/schedule/domain/entities/teaching_load.go
package entities

import (
	"errors"
	"time"

	"github.com/inf-sys-secretary-methodologist/inf-sys-secretary-methodist/internal/modules/schedule/domain"
)

// Domain validation errors returned by NewTeachingLoad.
var (
	ErrInvalidLoadReference = errors.New("teaching load references must be positive")
	ErrInvalidLoadPairs     = errors.New("pairs per week must be positive")
	ErrInvalidLoadWeekType  = errors.New("invalid week type")
)

// Repository-level sentinels for TeachingLoad lookups and constraints.
var (
	ErrTeachingLoadNotFound  = errors.New("teaching load not found")
	ErrTeachingLoadDuplicate = errors.New("teaching load for this group/discipline/type already exists")
)

// TeachingLoad is one planned teaching assignment: a group studies a discipline
// with a teacher for N pairs per week (on all/odd/even weeks). It is the source
// of truth the auto-scheduler expands into schedulable lessons.
type TeachingLoad struct {
	ID           int64           `json:"id"`
	SemesterID   int64           `json:"semester_id"`
	GroupID      int64           `json:"group_id"`
	DisciplineID int64           `json:"discipline_id"`
	TeacherID    int64           `json:"teacher_id"`
	LessonTypeID int64           `json:"lesson_type_id"`
	PairsPerWeek int             `json:"pairs_per_week"`
	WeekType     domain.WeekType `json:"week_type"`
	CreatedAt    time.Time       `json:"created_at"`
	UpdatedAt    time.Time       `json:"updated_at"`

	// Hydrated associations for read models (optional).
	Group      *StudentGroup `json:"group,omitempty"`
	Discipline *Discipline   `json:"discipline,omitempty"`
	LessonType *LessonType   `json:"lesson_type,omitempty"`
	Teacher    *TeacherInfo  `json:"teacher,omitempty"`
}

// NewTeachingLoad constructs a validated TeachingLoad. STUB — see GREEN commit.
func NewTeachingLoad(semesterID, groupID, disciplineID, teacherID, lessonTypeID int64, pairsPerWeek int, weekType domain.WeekType, now time.Time) (*TeachingLoad, error) {
	return &TeachingLoad{
		SemesterID:   semesterID,
		GroupID:      groupID,
		DisciplineID: disciplineID,
		TeacherID:    teacherID,
		LessonTypeID: lessonTypeID,
		PairsPerWeek: pairsPerWeek,
		WeekType:     weekType,
		CreatedAt:    now,
		UpdatedAt:    now,
	}, nil
}
